Feature support for async prefetch in inferno-router

jhsware opened the pull request:
All tests are passing for me.

This PR adds a property **loader** to \<Route\> which will be resolved prior to performing a navigation. This allows data to be fetched from an API and the result sent to the rendered component. The behaviour mimics the `loader` behaviour of React Router v6.

```
 <Route path="/" loader={async () => {} }
```

For SSR you can run `await resolveLoaders(...)` on your initial application and get the results of all the invoked loaders for a given location. This can then be passed to a \<StaticRouter loaderData={...} \> for server side rendering and then used during rehydration in the browser by passing the same data to \<BrowserRouter loaderData={...} \>.

This will greatly simplify SSR with Inferno which currently requires both patience and guts.

I have added some helper methods to abstract access to data in a way that looks familiar to those who are accustomed to the hooks of react-router v6. This has the benefit of allowing us to change implementation details without breaking applications.

Reference: https://reactrouter.com/en/main/route/loader

jhsware commented:
"renders on initial" is a faulty test, I will fix.

jhsware commented:
By removing async/await from browser bundled code and moving resolveLoader to inferno-server the inferno-router bundle returns to an acceptable size:
│ inferno-router         │ 39.77 kB/7.77 kB   │ 14.31 kB/4.37 kB          │ 31.76 kB/6.84 kB  │ 35.59 kB/7.47 kB  │ 14.16 kB/4.28 kB │

Original cjs prod: 12.8kB
New cjs prod with async loader support: 14.16 kB

jhsware commented:
TODO: Consider renaming **loaderData** to **initialData** to match initialEntries.
TODO: Pass params to async loader
TODO: Pass request to async loader

Havunen commented on packages/inferno-router/src/BrowserRouter.ts:
```diff
@@ -15,15 +16,16 @@ export interface IBrowserRouterProps {
+  constructor(props?: IBrowserRouterProps, context?: any) {
     super(props, context);
-    this.history = createBrowserHistory(props);
+    this.history = createBrowserHistory(/*props*/); // TODO: None of the props are defined in createBrowserHistory so skipping
```
What props the method have? Shouldnt we fix typing rather than change behavior?

jhsware replied:
The only property createBrowserHistory accepts is `window`, but we don't accept it as a prop to BrowserRouter. createBrowserHistory defaults `window = document.defaultView! ` so passing props here is a NOOP. Apparently one can pass another window object if one likes but I am not sure what use that is and what the side effects are.

We have two options:
1. remove props as above
2. add `window?: Window` to IBrowserRouterProps

What do you think makes more sense? NOTE: I didn't research this at all because it felt like out of scope, I just fixed the type error since it was a NOOP and assumed it was a copy/paste error.

Ref: https://github.com/remix-run/history/blob/main/packages/history/index.ts

Havunen replied:
Yeah based on your comments removing it makes sense. I can check it as well once I'm home.

jhsware replied:
@Havunen did you reach some conclusion on this? If we remove it, it will be easy enough to reintroduce if someone raises an issue.

Havunen replied:
yeah lets remove it

Havunen commented on packages/inferno-router/src/Route.ts:
```diff
@@ -87,18 +120,36 @@ class Route extends Component<Partial<IRouteProps>, RouteState> {
-    const { children, component, render } = this.props;
+    const { match, __loaderData__ } = this.state!;
+    
+    // QUESTION: Is there a better way to invoke this on/after first render?
```
Why this has to be in render method?

jhsware replied:
It shouldn't be there... I have now moved it to `componentDidMount` where it belongs 😎

jhsware commented:
Update: I have working tests with \<Switch\> but it needs some cleanup and more test cases.

jhsware commented:
@Havunen I have implemented the code in \<Switch \>
- refactored `function extractMatchFromChildren` (readability, size but adds one extra function call when resolving switch)
- TODO: preferably need more test cases
- TODO: there is room to DRY up Route and Switch
- TODO: I need to look at the timing to make sure we don't remove the current route before loader has been resolved
- TODO: also need to look at support for progress bar/loading state (could be released without)

jhsware commented:
I have added a demo app so we can visually check that routing behaves as intended. It appears that the outgoing route is removed before data has been returned. I need to investigate.

jhsware commented:
I have resolved the timing issue by giving Router responsibility for invoking loaders (I was thrown off by the location matching being done in each Route and placed the loader calls there first). This will also help implementing progress tracking. Broke some stuff but I am hoping it only requires some minor fixes.

jhsware commented:
All tests are passing and demo works as expected. I'll try to clean up the code, fix some naming inconsistencies and address the items in https://github.com/infernojs/inferno/pull/1621#issuecomment-1357329200

jhsware commented:
@Havunen I think we might have the feature set in place now. I believe I have implemented all the important features. Let me know if this makes sense. A request can be aborted, but I haven't implemented form submissions. It requires more work and form submission can be handled by the component without inferno-router being involved so I think we can live without it for starters.

I need to check the SSR-stuff once more because I updated resolveLoaders.ts, but it will be pretty straight forward.

jhsware commented:
@Havunen **I think the PR is complete now.** If you want to explore how it works you can: 
```
npm run build
cd demo/inferno-router-demo 
npm run dev
```
- requires Node 18 due to SSR-implementation
- supports client HMR and transpilation of SSR-bundle on start up using ParcelJS
- I have set an artificial delay to show that the old view stays until the new is ready

When you are happy, I will update the docs:
- https://www.infernojs.org/docs/api/inferno-router

I also intend to add SSR-templates which would make creating Inferno SSR-apps using Node 18 a breeze:
- https://github.com/jhsware/inferno-parcel-templates

Havunen commented on demo/inferno-router-demo/package.json:
```diff
@@ -0,0 +1,42 @@
+    "@parcel/transformer-sass": "^2.8.1",
+    "@parcel/transformer-typescript-types": "^2.8.1",
+    "@types/node": "^18.11.9",
+    "babel-plugin-inferno": "^6.5.0",
```
this example could use ts-plugin-inferno, no need for babel then

Havunen replied:
https://github.com/infernojs/ts-plugin-inferno

jhsware replied:
I looked at it but didn't understand how to configure ts-plugin-inferno for ParcelJS. Parcel uses SWC for transpilation of TS-code. If you have some pointers I would love to try it out.

Havunen replied:
Ah ok

Havunen commented on demo/inferno-router-demo/package.json:
```diff
@@ -0,0 +1,42 @@
+  "browserslist": "> 0.5%, last 2 versions, not dead",
+  "scripts": {
+    "dev": "npm-run-all -l --parallel  dev:*",
+    "dev:frontend": "FORCE_COLOR=1 parcel",
```
these scripts does not run on windows due to different syntax of environment variables than unix, please use cross-env

jhsware replied:
WILL FIX

jhsware replied:
Fixed in 65c7d77a

Havunen commented:
ts lint is reporting errors, can you check those please

Havunen commented:
demo app is also logging warning to console: 

> You are running production build of Inferno in development mode. Use dev:module entry point.

Havunen commented:
I noticed that if the user clicks page two in the demo app multiple times quickly and then clicks page one after it, it does not move to the page one but stays on page two. It might be a bug in the demo app or in the library

![image](https://user-images.githubusercontent.com/2021355/210258356-646384fb-2021-4c27-9b6d-bfea67ff0020.png)

Havunen commented on demo/inferno-router-demo/src/pages/PageTemplate.tsx:
```diff
@@ -0,0 +1,36 @@
+  children: any,
+}
+
+interface IState {
```
dummy state interface

jhsware replied:
WILL FIX

jhsware replied:
Fixed in e746f08cf8

Havunen commented on demo/inferno-router-demo/src/server.ts:
```diff
@@ -0,0 +1,164 @@
+import koa from 'koa';; // koa@2
```
double semicolon and unnecessary imports

jhsware replied:
WILL FIX

jhsware replied:
Fixed in 8e292ba6

Havunen commented on demo/inferno-router-demo/tsconfig.json:
```diff
@@ -0,0 +1,18 @@
+{
+  "compilerOptions": {
+    "esModuleInterop": true,
```
it would be nice to have these 2 settings false: 

```
    "esModuleInterop": false,
    "isolatedModules": false,
``` 
but then code needs some cleaning

jhsware replied:
WILL INVESTIGATE

- I believe I had an issue with third-party libraries, but I will give it a try.

jhsware replied:
Fixed in 12d29a3

Havunen commented on packages/inferno-router/__tests__/loaderWithSwitch.spec.jsx:
```diff
@@ -0,0 +1,298 @@
+    const link = container.querySelector('#link');
+    link.click();
+
+    // TODO: I believe this test gives a false pass, in a demo, the page switches too early
```
is this still an issue?

jhsware replied:
The problem in the code is fixed, but the test passed before I implemented the fix. I'll see if I can make the test more accurate.

jhsware replied:
Fixed in 625c5e0

Havunen commented on packages/inferno-router/__tests__/loaderOnRoute.spec.jsx:
```diff
@@ -0,0 +1,586 @@
+    link.firstChild.click();
+
+    // Wait until async loader has completed
+    await waitForRerender();
```
inferno has a method called "rerender" which can be called to flush all pending renders, then there is no need to wait

jhsware replied:
This is also waiting for the async loader to complete. I will look att combining the two.

jhsware replied:
Fixed in 195f4686c (note: all the awaits are needed too)

Havunen commented on packages/inferno-router/src/resolveLoaders.ts:
```diff
@@ -0,0 +1,198 @@
+}
+
+// Optionally pass base param during SSR to get fully qualified request URI passed to loader in request param
+export function traverseLoaders(location: string, tree: any, base?: string, parentIsSwitch = false): TLoaderEntry[] {
```
hmm, so this method is exposed to public and contains a boolean information about parent node being a switch? `parentIsSwitch` I would like to get that flag removed if possible

Havunen replied:
This method could be written using a stack of nodes to avoid recursion and avoid exposing switch parameter externally

Havunen replied:
a minimum change could be to have a separate method for traverseLoaders which has parentIsSwitch and call that internally here and dont expose it

jhsware replied:
WILL FIX: I'll go for exposing a public method

jhsware replied:
Fixed in 0843db8fcb

Havunen commented on packages/inferno-router/src/resolveLoaders.ts:
```diff
@@ -0,0 +1,198 @@
+    const entries = tree.reduce((res, node) => {
+      if (parentIsSwitch && hasMatch) return res;
+
+      const outpArr = traverseLoaders(location, node, base, node?.type === Switch);
```
Switch is exposed and not sealed, so one can implement their own switch, I believe we should check for inheritance also here

jhsware replied:
WILL FIX

jhsware replied:
Fixed in b79c8f9e169

Havunen replied:
Here is still === Switch comparison

jhsware replied:
Fixed in  e7f33a97d (and b79c8f9)

Havunen commented on demo/inferno-router-demo/src/server.ts:
```diff
@@ -0,0 +1,164 @@
+  </body>
+</html>  
+`
+  return html;
```
unreachable code

jhsware replied:
WILL FIX

Havunen commented on packages/inferno-router/src/resolveLoaders.ts:
```diff
@@ -0,0 +1,198 @@
+import type { TLoaderData, TLoaderProps } from "./Router";
+import { Switch } from "./Switch";
+
+export function resolveLoaders(loaderEntries: TLoaderEntry[]): Promise<Record<string, TLoaderData>> {
```
IMO this could maybe be async method then it could await on Promise.all  call to make it slightly cleaner

jhsware replied:
I tried that from the outset, but this added async/await code to the transpiled bundle and it grows substantially. So I replaced all async/await with promise/then.

ref: https://github.com/infernojs/inferno/pull/1621#issuecomment-1357294390

Havunen commented on packages/inferno-router/src/resolveLoaders.ts:
```diff
@@ -0,0 +1,198 @@
+function resolveEntry(path, params, request, loader): Promise<any> {
+  return loader({ params, request })
+    .then((res) => [path, { res }])
+    .catch((err) => [path, { err }]);
```
so err is `Error`  that could be typed

jhsware replied:
Both `err` and `res` comes from the implemented loader so we don't really know what it looks like. The type won't be passed to the consuming code anyway. I will look at adding typing for the helper methods used to access the data:

WILL INVESTIGATE

ref: [packages/inferno-router/src/helpers.ts](https://github.com/infernojs/inferno/pull/1621/files#diff-ca73e06743e22d5dffd9a3baa6f39f1a0c3c49be0317cf6f46f235d3c7ecce62)

jhsware replied:
Fixed in e6499632ed

Havunen commented on packages/inferno-router/src/Router.ts:
```diff
@@ -1,46 +1,89 @@
+
+export type TLoaderData = {
+  res?: Response;
+  err?: any;
```
Error type?

jhsware replied:
This is unknown at this point since the actual implementation of the error is up to the developer (could be anything from a simple error message string, a JSON-object or an actual error object).

Ref: https://github.com/infernojs/inferno/pull/1621#discussion_r1060147378

jhsware commented:
> demo app is also logging warning to console:
> 
> > You are running production build of Inferno in development mode. Use dev:module entry point.

I believe ParcelJS is consuming the wrong bundle. I will see if I can solve it without requiring a custom resolver.

WILL INVESTIGATE

jhsware commented:
@Havunen thank's for all the awesome feedback. I will dig deeper over the next week.

jhsware commented:
> > demo app is also logging warning to console:
> > > You are running production build of Inferno in development mode. Use dev:module entry point.
> 
> I believe ParcelJS is consuming the wrong bundle. I will see if I can solve it without requiring a custom resolver.
> 
> WILL INVESTIGATE

I solved this by using a custom resolver, but it is simple enough and I believe we need to provide one to support the dev:module entry. I will improve on the resolver if we want to publish it, for the demo use case it is hard coded. We probably want to move it to it's own repos too.

Fixed in d5791987b

jhsware commented:
> Doesn't parcel have any "alias" support like rollup or webpack? I think this could be located in its own repository as we don't have anything bundling / packaging related in infernojs repo.

They do, but it won't be production/development aware. Perhaps okay to stick to development in the demo though.

I agree that the resolver should like in it's own repos. So I'll:
- use alias to dev versions of packages for demos
- develop the proper resolver in a package I own and if it turns out to be useful we move it to Inferno-control

jhsware commented:
I believe I have fixed all the outstanding issues that have been raised in this discussion. I went through my comments and I have created a PR for the website.

Features from react-router I would like to add but probably shouldn't stop us from releasing this:
TODO: react-router automatically calls .json() if needed which allows more compact loader functions
TODO: react-router supports progress reporting

Havunen commented on packages/inferno-router/README.md:
```diff
@@ -21,9 +21,11 @@ Features added from react-router@5:
+- download progress support
 - form submission
+- redirect support
+- not exposing headers, type or status code to render method
```
Hmm, why should we follow react router implementation?

I would change wording to just mention following features are not supported yet. 

But is it possible for application side to provide its own settings to fetch, thus implementing those are possible at application level?

jhsware replied:
> Hmm, why should we follow react router implementation?
My thoughts were that it is a react-router port from the outset so we should try to keep it compatible where we can.

> 
> I would change wording to just mention following features are not supported yet.
WILL FIX

> 
> But is it possible for application side to provide its own settings to fetch, thus implementing those are possible at application level?
Yes, the application is responsible for implementing the fetch call (or whatever library they want to use) so they have full control. Note: adding support for exposing response headers, form submission and redirect support will be a non-breaking change from what I can tell.

jhsware commented:
Rebased on master

Havunen commented on packages/inferno-server/__tests__/loaderOnRoute.spec.server.jsx:
```diff
@@ -0,0 +1,49 @@
+import { render } from 'inferno';
```
could this test be converted to TSX please :)

Havunen commented on packages/inferno-router/__tests__/BrowserRouter.spec.tsx:
```diff
@@ -26,12 +27,13 @@ describe('BrowserRouter (jsx)', () => {
 
     // browser only?
-    expect(console.error.calls.argsFor(0)[0]).toContain('<BrowserRouter> ignores the history prop');
+    expect((console.error as any).calls.argsFor(0)[0]).toContain('<BrowserRouter> ignores the history prop');
```
putting the result of spyOn(console, 'error'); to variable corrects the typing so "any" is not needed

jhsware replied:
Fixed in 54581c76

Havunen reviewed (approved):
build and tests should pass, otherwise LGTM! :)

jhsware commented:
@Havunen tests are passing and I fixed a bunch of linting errors too.
